File: crm-portal-ts/src/templating.ts
import { Request } from 'express';
import { Jinja2Templates } from 'jinja2-express';
import { UserRole } from './models/UserRole'; // Assuming UserRole is defined in a separate file
import { uiService } from './services/ui_service';

// Preconfigure Jinja2 environment to avoid DeprecationWarning
const env = new Jinja2Templates({
    loader: new jinja2.FileSystemLoader(__dirname + '/templates'),
    extensions: ['jinja2.ext.do']
});

const templates = env;

async function render(
    request: Request,
    name: string,
    context?: Record<string, any>,
    status_code?: number
): Promise<Response> {
    const ctx = { ...context };
    const u = request.state.portalUser;
    ctx['portal_user'] = u;
    const c = request.state.clientUser;
    ctx['client_user'] = c;

    let cachedNavItems: UserRole[] | undefined;
    if (u) {
        if (cachedNavItems !== undefined) {
            ctx['nav_items'] = cachedNavItems;
            ctx['nav_groups'] = groupedVisibleNav(cachedNavItems);
        } else {
            const db = await dbManager.getSessionLocal();
            try {
                const items = await visible_nav_items(db, u.role);
                ctx['nav_items'] = items;
                ctx['nav_groups'] = groupedVisibleNav(items);
            } finally {
                db.close();
            }
        }
        const r = u.role;
        ctx['can_write_crm'] = r === UserRole.admin || r === UserRole.manager;
        ctx['can_write_helpdesk'] = r === UserRole.admin || r === UserRole.manager || r === UserRole.service;
        ctx['can_write_messages'] = r === UserRole.admin || r === UserRole.manager || r === UserRole.service;
        ctx['can_mutate'] = ctx['can_write_crm'];
        ctx['visible_menu_keys'] = request.state.visibleMenuKeys ?? new Set();
    } else {
        ctx['nav_items'] = [];
        ctx['nav_groups'] = [];
        ctx['can_write_crm'] = false;
        ctx['can_write_helpdesk'] = false;
        ctx['can_write_messages'] = false;
        ctx['can_mutate'] = false;
        ctx['visible_menu_keys'] = new Set();
    }
    ctx['title'] = '';
    ctx['app_name'] = APP_DISPLAY_NAME;

    // UI Plugin Integration
    ctx['ui'] = uiService.getThemeAssets(request);
    ctx['breadcrumbs'] = uiService.getBreadcrumb(request, ctx['title']);

    const resp = templates.renderTemplate(request, name, ctx);
    if (status_code !== undefined) {
        resp.status(status_code);
    }
    return resp;
}

File: app/templating.py
from fastapi import Request
from fastapi.templating import Jinja2Templates
import jinja2

from app.config import APP_DISPLAY_NAME, BASE_DIR
from app.database import db_manager
from app.models import UserRole
from app.nav_access import grouped_visible_nav, visible_nav_items
from app.services.ui_service import ui_service

# Preconfigure Jinja2 environment to avoid DeprecationWarning
env = jinja2.Environment(
    loader=jinja2.FileSystemLoader(str(BASE_DIR / "templates")),
    extensions=["jinja2.ext.do"]
)

templates = Jinja2Templates(env=env)


def render(
    request: Request,
    name: str,
    context: dict | None = None,
    *,
    status_code: int | None = None,
):
    """Render szablonu z kontekstem nawigacji i flagą zapisu (rola view = tylko odczyt)."""
    ctx = dict(context or {})
    u = getattr(request.state, "portal_user", None)
    ctx["portal_user"] = u
    c = getattr(request.state, "client_user", None)
    ctx["client_user"] = c

    cached = getattr(request.state, "nav_items_for_template", None)
    if u:
        if cached is not None:
            ctx["nav_items"] = cached
            ctx["nav_groups"] = grouped_visible_nav(cached)
        else:
            db = db_manager.SessionLocal()
            try:
                items = visible_nav_items(db, u.role)
                ctx["nav_items"] = items
                ctx["nav_groups"] = grouped_visible_nav(items)
            finally:
                db.close()
        r = u.role
        ctx["can_write_crm"] = r in (UserRole.admin, UserRole.manager)
        ctx["can_write_helpdesk"] = r in (UserRole.admin, UserRole.manager, UserRole.service)
        ctx["can_write_messages"] = r in (UserRole.admin, UserRole.manager, UserRole.service)
        ctx["can_mutate"] = ctx["can_write_crm"]
        ctx["visible_menu_keys"] = getattr(request.state, "visible_menu_keys", set())
    else:
        ctx["nav_items"] = []
        ctx["nav_groups"] = []
        ctx["can_write_crm"] = False
        ctx["can_write_helpdesk"] = False
        ctx["can_write_messages"] = False
        ctx["can_mutate"] = False
        ctx["visible_menu_keys"] = set()
    ctx.setdefault("title", "")
    ctx.setdefault("app_name", APP_DISPLAY_NAME)
    
    # UI Plugin Integration
    ctx["ui"] = ui_service.get_theme_assets(request)
    ctx["breadcrumbs"] = ui_service.get_breadcrumb(request, ctx["title"])
    
    resp = templates.TemplateResponse(request, name, ctx)
    if status_code is not None:
        resp.status_code = status_code
    return resp
